daemons/execd/pacemaker-execd.c

```diff
@@ -32,11 +32,13 @@ int lrmd_call_id = 0;
 #ifdef PCMK__COMPILE_REMOTE
 /* whether shutdown request has been sent */
 static sig_atomic_t shutting_down = FALSE;
+static gboolean have_shutdown_ack = FALSE;
```

Comment:
Thinking about it some more, we should be able to fast-exit whenever shutting_down is true -- there's a 20-second timer that would keep us from waiting forever if the connection drops before the ack has arrived, but there's no need to wait for the timer in that case

As an aside, I just realized shutting_down doesn't need to be sig_atomic_t since it's set by a mainloop "signal handler", not an actual interrupt-driven signal handler

Reply:
ahh, so we can exit from lrmd_remote_connection_destroy() if shutting-down is true? I suppose waiting for the ack is a bit pointless then anyway.

I was testing this originally by putting an exit() into the main server code that sends the ACK :)

Reply:
oh. If if doesn't go in  lrmd_remote_connection_destroy() then I'm not sure where it would.

Reply:
I reversed lrmd_tls_server_destroy() and  lrmd_remote_connection_destroy() in my mind, after just getting them straight :)

Reply:
It sounds like a rename is definitely needed then! I've included that in this latest patch, see what you think to the names, it's quite hard distinguishing between the various connections in there!